src/stream/src/common/table/state_table.rs

```diff
@@ -938,6 +937,34 @@ where
             .map(get_second))
     }
 
+    async fn iter_all_pk_and_val_with_pk_range(
+        &self,
+        pk_range: &(Bound<impl Row>, Bound<impl Row>),
+        prefetch_options: PrefetchOptions,
+    ) -> StreamExecutorResult<RowStreamWithPk<'_, S, SD>> {
+        let mut vec = Vec::with_capacity(self.vnodes.count_ones());
+        for vnode in self.vnodes.iter_vnodes() {
+            vec.push(
+                self.iter_key_and_val_with_pk_range(pk_range, vnode, prefetch_options)
+                    .await?,
+            )
+        }
+        let pinned_iter: Vec<_> = vec.into_iter().map(Box::pin).collect_vec();
+        let iter = merge_sort(pinned_iter);
+        Ok(iter)
+    }
+
+    pub async fn iter_all_with_pk_range(
```

Comment:
IIUC, the reason why we need this interface is because backfill needs to do a full table scan on the relevant vnodes and gets back the results sorted solely by PK. Although the previous backfill implementation without arrangment backfill did the same thing with the storage table interface, I realize that in theory backfill doesn't require the results fetched in the snapshot phase to be given with PK order in the table. Instead, it is still correct if backfill gets the result in VNODE + PK order because both input/output of backfill is sharded by vnode and a row can only map to one vnode in backfill. Therefore, I want to bring up a discussion here.

I prefer sorting by VNODE + PK because:
- It can be more CPU efficient and memory efficient since we don't need the unneccesssary merge sort.
- We have better read locality since data stored in storage are sorted by vnode + PK.
- It still maintains the assumption that each call to the state table only access one vnode and no new interfaces are needed.
- It aligns with the backfill internal state better, which is a per vnode state (Schema=| vnode | pk | backfill_finished |)

One simple way to adopt sorting by VNODE + PK in the snapshot phase is to scan sequentially on the vnodes backfill is responsible for and the handling of the buffered stream messages are a littble bit different:
- Skip a row if it belongs to a not-yet-scanned vnode or a scanning vnode with smaller current_pos
- Emit a row to downstream if it belongs to a fully-scanned vnode.
- Emit a row to downstream if it belongs to a scanning vnode with larger current_pos

One may think that this may affect the parallelsim of processing since we scan one vnode at a time. But the paralleism of the backfill executor is still preserved and each of them scans the vnodes in parallel so I think it is accetable.

Reply:
I think this algorithm is correct and feasible, though I am sure how many benefits it could bring since we need to calculate the vnode for each upstream chunk and check whether it is a not-yet-scanned vnode later. My biggest concern is whether it could result in some data skewing for the downstream during backfilling, since each time we output vnode data individually. It seems possible that the upstream distribution key has some correlation with the distribution key. The downstream could be the user system when the backfill executor is used by creating sink scenarios.

Reply:
Fair enough. There are indeed two independent issues here:
1. Vnode calculation for upstream rows
2. Scanning vnode sequentially causing data skewing in dowstream during mv backfilling.

For 1, I thought it is fine since per row vnode calculation happens almost in all stateful executors to calculate the bloom filter key before reading from storage.
For 2, I think it can be resolved by not scanning vnodes sequentially but wrapping the iterators for each vnode in a FutureUnordered. In this case, we can still avoid merge sort.

Reply:
That's true. Maintaining each vnode corresponding `current_pos` in memory just like how we store it in the state table and iterate each vnode scanning iterator in a round-robin way sounds good.

Reply:
Let me see if I got it:
1. Each parallelism has set of vnodes.
2. Within each parallelism scan vnodes in round-robin fashion to avoid data skew.
3. Backfill executor will no longer have a single `current_pos`, but `(vnode, current_pos)` to track backfill progress per vnode.
4. Step 2 can be accomplished with `FuturesUnordered`: https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html.

---

> For 1, I thought it is fine since per row vnode calculation happens almost in all stateful executors to calculate the bloom filter key before reading from storage.

Actually we manage to avoid this cost so far for backfill executor. This is because we currently update current_pos for ALL vnodes.
https://github.com/risingwavelabs/risingwave/blob/79efa32cc635d780a61ac109237600e5076775b5/src/stream/src/executor/backfill.rs#L574-L582

In the code snippet above, `current_partial_state` contains the current_pos. Then we just write this `current_pos` for each vnode.

This means that vnodes are not computed, but instead fetched from vnodes of table.

I guess this logic no longer holds with the new algorithm. We need to maintain independent backfill progress PER vnode instead. That means when we process each row, we no longer can check against a single `currrent_pos` to decide how to handle it. Instead, we need to check against the corresponding `(vnode, current_pos)` pair. This also means we will need to compute vnode of each row which will be more expensive.

This change may be inevitable if we also wish to support recovery by persisting state. Previously we had the issue of not knowing how to recover with multiple backfill executor scale-in, because backfill logic assumes that all vnode are processed to the same `current_pos`.
With this change, different vnode now can have various stage of progress.

So I think this change makes a lot of sense.

---

If we are concerned about performance, I guess we can have a purely in-memory backfill executor, which does not persist state at all, and let user decide if they wish to use it.

This in-memory backfill executor can be used as benchmarking baseline against our other backfill executors.

Reply:
Indeed, this can make scale-in or migration of creating mv possible. I think their performance is basically the same. I think now we have good reason to use this algorithm for the arrangement backfill. For the no-shuffle backfill, since it uses a storage table that doesn't have an interface to scan a specific vnode, we can keep it as usual.